Add EC2 autograder features

evanyeyeye opened the pull request:
## Description
This PR enables the use and configuration of EC2 autograding instances via Autolab. 

Documentation will be added in a future PR. 

## How Has This Been Tested?
- Verified that requests to Tango properly utilize the EC2 vmms.
- Verified that configured defaults work correctly.
- Verified that custom access credentials can be passed to Tango on a per job basis.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] I have run rubocop and erblint for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [x] My change requires a change to the documentation, which is located at [Autolab Docs](https://docs.autolabproject.com/)
- [ ] I have updated the documentation accordingly, included in this PR

20wildmanj committed:
testing ec2, add support on autolab side

20wildmanj committed:
add support for EC2 settings in Autolab

20wildmanj committed:
fix issue with autograder initialization

20wildmanj committed:
add new option on Tango side to explicitly check for ec2 vmms usage, and disable access key fields if not enabled

20wildmanj committed:
add ec2ssh to prod config

20wildmanj committed:
temporary changes to help with testing autolab 3

20wildmanj committed:
Revert "temporary changes to help with testing autolab 3"

This reverts commit fc35acec7b66926efe71970ff85b9e0f4032da48.

KesterTan committed:
Merge conflicts

KesterTan committed:
Merge branch 'master' into ec2-testing

evanyeyeye committed:
Minor touchups

evanyeyeye committed:
Merge branch 'ec2-testing' of github.com:autolab/Autolab into ec2-testing

Futurecomputerfixer committed:
ec2-dropdown

Futurecomputerfixer committed:
fix linting

Futurecomputerfixer committed:
fix linting #2

Futurecomputerfixer committed:
try fix error parsing for scoreboard

KesterTan committed:
Merge branch 'master' into ec2-testing

KesterTan reviewed (approved):
LGTM -> tested on https://dev.autolabproject.com/

Futurecomputerfixer reviewed (changes requested):
Security (secrets handling)

Problem: access_key / access_key_id are stored and passed around as plaintext. They must be encrypted at rest, scrubbed from logs, and not echoed in the UI.

Changes:

Prefer IAM roles over keys

File: app/helpers/assessment_autograde_core.rb

In the EC2 block, if use_access_key is false, omit accessKey/accessKeyId entirely from job_properties (don’t send empty strings).

Encrypt at rest (follow GitHub integration precedent)

Add encrypted columns mirroring github_integrations.access_token_ciphertext.

New migration (replace plaintext columns introduced in this PR):

Files:

Replace db/migrate/20241205233214_add_ec2_ssh_fields_to_autograders.rb with one that adds:

access_key_ciphertext :text

access_key_id_ciphertext :text

instance_type :string, default: "" (kept)

Drop the plaintext access_key/access_key_id columns if already migrated.

Model:

File: app/models/autograder.rb

Use the same encryption library already used (e.g., Lockbox) to define:

encrypts :access_key
encrypts :access_key_id


where those virtual attributes map to *_ciphertext.

Gate presence validations with use_access_key (see “Validations” section).

Scrub from logs & params filtering

File: config/initializers/filter_parameter_logging.rb

Add: Rails.application.config.filter_parameters += %i[access_key access_key_id]

File: app/controllers/autograders_controller.rb

Ensure we never flash or log raw values.

UI should not echo secrets

File: app/views/autograders/_ec2_settings.html.erb

Change inputs to password-style and avoid prepopulating:

<%= f.password_field :access_key, display_name: "Access Key", value: nil, autocomplete: "new-password" %>
<%= f.text_field :access_key_id, display_name: "Access Key ID", value: nil, autocomplete: "off" %>


Add a small hint: “Leave blank to keep existing secret.”

Strong params & controller flow

Problem: Params are permitted in two calls; :use_access_key isn’t included; and empty-string secrets are sent downstream.

Changes:

File: app/controllers/autograders_controller.rb

Unify strong params and include all fields:

def autograder_params
  params.require(:autograder).permit(
    :autograde_timeout, :autograde_image, :release_score,
    :use_access_key, :access_key, :access_key_id, :instance_type
  )
end


In create defaults block, avoid initializing secrets to "". Prefer nil:

a.access_key_id = nil
a.access_key = nil
a.instance_type = "t3.micro" # see validations below


In update/create, if use_access_key is false, ignore any provided access_key/access_key_id (set to nil before save).

Validations & defensive checks

Problem: instance_type is free-form; secrets aren’t validated when enabled.

Changes:

File: app/models/autograder.rb

Add allow-list for instance types shown in the UI:

INSTANCE_TYPES = %w[
  t2.nano t2.micro t2.small t2.medium t2.large t2.xlarge t2.2xlarge
  t3.micro t3.small t3.medium t3.large
  c5.large c5.xlarge c5.2xlarge
  r5.large r5.xlarge r5.2xlarge
].freeze

validates :instance_type, inclusion: { in: INSTANCE_TYPES }, allow_blank: true


Gate secrets validation behind the toggle:

with_options if: :use_access_key do
  validates :access_key_id, presence: true, format: { with: /\A[A-Z0-9]{16,24}\z/, message: "looks invalid" }
  validates :access_key, presence: true
end


Add bounded timeout validation if not already present:

validates :autograde_timeout, inclusion: { in: 10..900 }

Tango payload assembly

Problem: Empty strings are sent; json built twice; feature flag should be respected server-side.

Changes:

File: app/helpers/assessment_autograde_core.rb

Build a Ruby hash first, then JSON once at the end; respect config.x.ec2_ssh and use_access_key:

job_properties = {
  "image" => @autograde_prop.autograde_image,
  "timeout" => @autograde_prop.autograde_timeout,
  "callback_url" => callback_url,
  "jobName" => job_name,
  "disable_network" => assessment.disable_network
}

if Rails.configuration.x.ec2_ssh
  job_properties["ec2Vmms"] = true
  job_properties["instanceType"] = @autograde_prop.instance_type.presence || "t3.micro"

  if @autograde_prop.use_access_key?
    job_properties["accessKey"]   = @autograde_prop.access_key
    job_properties["accessKeyId"] = @autograde_prop.access_key_id
  end
end

job_properties = job_properties.to_json


This ensures keys are omitted entirely when not used.

Frontend (JS & form UX)

Problem: Minor JS bug; console noise; better UX for disabled/secret fields.

Changes:

File: app/assets/javascripts/autograder.js

Fix .val() misuse and remove console log:

;(function() {
  $(document).ready(function () {
    function access_key_callback() {
      const checked = $(this).prop('checked');
      const $access_key_field = $('#autograder_access_key');
      const $access_key_id_field = $('#autograder_access_key_id');
      $access_key_field.prop('disabled', !checked);
      $access_key_id_field.prop('disabled', !checked);
      if (!checked) {
        $access_key_field.val('');
        $access_key_id_field.val('');
      }
    }

    $('#autograder_use_access_key').on('change', access_key_callback);
    access_key_callback.call($('#autograder_use_access_key'));

    // Optional: keep dropdown UI tweaks, but drop console.log
    $('#autograder_instance_type option').hover(
      function() { $(this).addClass('highlighted-option'); },
      function() { $(this).removeClass('highlighted-option'); }
    );
  });
})();


File: app/views/autograders/_ec2_settings.html.erb

Add pattern to Access Key ID (light client-side feedback):

<%= f.text_field :access_key_id, display_name: "Access Key ID",
                 value: nil, autocomplete: "off",
                 pattern: "[A-Z0-9]{16,24}",
                 title: "Looks like an invalid AWS Access Key ID" %>

Feature flag correctness

Problem: UI is gated by config.x.ec2_ssh, but server must guard too.

Changes:

Files:

config/environments/development.rb – OK to be true.

config/environments/production.rb.template – defaults to false (good).

Server guard:

File: app/helpers/assessment_autograde_core.rb

Ensure the EC2 payload branch only runs if Rails.configuration.x.ec2_ssh is true (already applied above).

Migrations / schema churn

Problem: schema.rb shows charset/timestamp engine churn unrelated to this PR; plaintext columns introduced for secrets.

Changes:

Files: db/schema.rb

Regenerate schema after replacing the plaintext secret columns with encrypted *_ciphertext. Avoid unrelated charset/timestamp diffs (ensure you run the same DB engine/version locally as CI).

Files: db/migrate/20241205233214_add_ec2_ssh_fields_to_autograders.rb, db/migrate/20241211042124_add_use_access_key_to_autograder.rb

Update to remove plaintext columns and add encrypted ones (see “Security” section). Keep use_access_key:boolean migration.

Tests

Problem: No tests cover new payload assembly and validations.

Add:

File: spec/helpers/assessment_autograde_core_spec.rb

Cases:

ec2_ssh=false → no ec2Vmms/instanceType/keys in payload.

ec2_ssh=true, use_access_key=false → ec2Vmms & instanceType present, no keys.

ec2_ssh=true, use_access_key=true → keys present.

File: spec/models/autograder_spec.rb

Validates inclusion for instance_type.

Validates secrets presence only when use_access_key=true.

KesterTan committed:
Merge branch 'master' into ec2-testing

KesterTan committed:
encrypt access keys

KesterTan committed:
linting

Futurecomputerfixer committed:
raise exception if ece2 type dont exist

Futurecomputerfixer committed:
linting

anthony-yip committed:
Slightly more robust handling of job completion status (#2301)

KesterTan committed:
encrypt access keys

jhs-panda committed:
Add options for G-instance types (#2311)

Add g4dn.xlarge, g4dn.2xlarge, and g5.xlarge instance type options

anthony-yip committed:
Allow lint check to pass (#2306)

merged if-else branches

NicholasMy committed:
Resolve security concerns from upstream issue #2313 (#2317)

(cherry picked from commit 9ccdfbbfd910c1ad900bcbea576099330f7492be)

coder6583 committed:
Permission Model + Image Builder (#2325)

* implemented a file system permission cleaner and hooked it up in each web api function to enforce it

* auto infer group name from path

* modify ssh

* fix to test on gamma

* Implement SSH key management and Unix user/group system

- Add SshKey model with validation and fingerprinting
- Implement UnixGroupManager service for user/group management
- Create Unix users/groups automatically on course/staff creation
- Add SSH key provisioning to authorized_keys
- Implement FilesystemEnforcer with group-based permissions
- Add SSH key management UI (restricted to staff/instructors)
- Fix require paths for services in controllers
- Add services directory to autoload paths
- Create bootstrap script for existing courses
- Add host system user setup scripts
- Support BusyBox useradd (use -p '*' instead of --disabled-password)
- Add comprehensive setup and testing scripts

* testing

* add a new daemon container

* fix bugs

* fix bugs ++

* add hint for ssh key creation

* add symlink

* testing ec2, add support on autolab side

* Filesystem mapping + EC2 autograder changes

* fix group

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* add support for EC2 settings in Autolab

* fix issue with autograder initialization

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* fix groupa

* editing ssh commands

* added service user to classes

* attempt to fix course permissions

* unix file system scanner

* install assessment

* Revert "install assessment"

This reverts commit 224f4ee3c9fd0bbd5b9f489fda3ff877e5707ade.

* Revert "unix file system scanner"

This reverts commit 73326f8a2f29aba760f16042270021c691b395cc.

* Revert "attempt to fix course permissions"

This reverts commit 18d9fa478d8c6deef76b513fe6b1a1234fa52ab4.

* make user9999 always an instructor

* test adding user9999 to all groups

* remove integration bundle

* fix user groups

* remove check of local membership

* fix script to have both local and group

* tranfer ownership script

* CAs should not be able to ssh

* Add clarification about course unix group inclusions.

* modify ssh commands

* add provision and deprovision of keys

* prevent group add error

* fix typo in ssh commands

* fixed ca and instructor ssh perms

* sponsors and contact us page

* try to add both app and user9999

* list dir to unix ops

* fix unix ops delegation

* change delegate parsing

* mkdir over to unixops

* formatting

* should delegate during perm error

* restrict dump yaml

* yaml read file delegate to unix ops

* is file delagated to unix ops

* dump yaml to delegate

* conversion of pathname

* load config to delegate

* guard against empty file read from delegate

* try reload config file with fallback

* handin directory exists delegate

* bulk email and delegation of read writeup

* delegate submission to unix ops

* unable to read files and submit files

* Submission unviewable

* check if use_access_key exists

* basic controller done, before tango integration

* test modified symlink

* try test modified symlink fix

* update permissions when ssh key changes/user role changes

* add cleanup step for symlinks

* try fix symlink

* add log statement to debug

* modify unix_ops_daemon

* test file transfer change

* undo prev changes

* try new change

* revert changes

* test new change

* try new fix

* try new fix

* try new fix

* fix circular dependency

* try to fix for all except admin

* remove user's directory once it has no courses

* handle cleanup for deleted users

* reverse unix group deletion

* change symbolic link ownership

* try to fix cd update issue

* try to fix cd update issue

* fix circular looping

* try to fix circular looping

* fix file access

* undo file access changes

* creating container image works

* table UI almost done

* refresh status implemented

* make rails DB source of truth

* refresh all status when reloading the page/ added refresh all button

* access key

* attempt to fix file perms

* instance type guards

* fixed migrations

* logs get updated

* format error

* create parents if it does not exist

* course creation delegate

* rb sys delegate

* list dir paths to delegate

* fallback to list_dir for path exists

* fix for view submission

* fix view for submissions controller

* delegate reading to unix

* more archive read paths to be delagted

* fix file manager

* delete feedback and path access fallbacks

* delegate delete and move to unix ops

* add delete and move path to unix ops

* only show ssh part if instructor

* try modifying users list breadcrumb

* modify "Users List" view to look the same as when it has a link

* modify symlinks to handle admin properly

* can view log now

* additional details for uploading job

* fallback for tango upload

* allow for non empty feedback without autograder

* autograding works with ec2docker now!

* manage autograder images on manage course

* debugging for autograding feedback

* delegated write of feedback to unix ops

* view feedback delegate

* add description clarification

* more unix ops delegation and documentation

* fix autograder upload errors

* debugging for saving autograder

* autograder length increased

* delegate list assessment dirs

* upload uses delegate

* disallow duplicate keys across all users

* distinguish between duplicates for current user and copies from other users

* Filter out non-ready container images in autograder settings and add … (#2322)

Filter out non-ready container images in autograder settings and add dockerfile_contents column

* remove ssh access for user once account deleted (keep unix account on system)

* revert change

* added admin container images controller + routes

* add logic to wipe ssh access once account deleted

* added template for managing images

* added template for viewing build log

* added template for new image page

* try to prevent unix account creation for tas

* created admin pages

* new image form works now

* can manage images with correct permissions now

* public images have unique names

* add public_template_id to container_images

* can select public templates now

* admin controller names don't overlap anymore

* admin controller names don't overlap anymore, routes

* passes base docker image tag and uri to tango

* building image works without public template now

* authorized false by default

* only show container images that belong to the course

* gemfile dockerfile update

* added button to the manage public templates page

* Fixed polling issue with ECR builder (#2323)

* export delegate

* order images in newly created order

* add unique unix_user column to users table

* fixed public_template.image_uri

* attempt to fix non-unique ssh login issue

* instructors can choose public images now

* fixed bug with join

* push unix_user database migration file

* load dir to tar to delegate

* lint fix

* course directory name

* change ruby version

* clean up lint issues

* clean up lint issues

* changed production.rb template to add ec2_docker

* added helper text to new docker image form

* increase client max body in nginx

* remove name index

* debugging for admin ssh key

* added prefix to unix group

* remove t2 nano

---------

Co-authored-by: Haoyu Yang <haoyuy@andrew.cmu.edu>
Co-authored-by: Joey Wildman <josephwildman88@gmail.com>
Co-authored-by: Kester <kestertan040@gmail.com>
Co-authored-by: Joy Song <joys@andrew.cmu.edu>
Co-authored-by: Giancarlo Umberto Ambrosino <129298577+Giabbi@users.noreply.github.com>
Co-authored-by: Ubuntu <ubuntu@ip-172-31-84-53.ec2.internal>

KesterTan committed:
Fix AWS setup and SSH (#2326)
